java/src/main/java/com/google/appengine/tools/txn/TxnWrapper.java

```diff
@@ -0,0 +1,100 @@
+package com.google.appengine.tools.txn;
+
+
+import com.google.appengine.tools.pipeline.impl.backend.PipelineTaskQueue;
+import com.google.cloud.datastore.Transaction;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Multimap;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.java.Log;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+
+/**
+ * Transaction wrapper class that aims to mimic cross-services transactions. In this case datastore-cloud tasks.
+ */
+@Log
+@RequiredArgsConstructor(staticName = "of")
+public class TxnWrapper {
+
+  @NonNull
+  @Getter // should only be accessed when adding stuff to the txn
+  Transaction dsTransaction;
+
+  final PipelineTaskQueue taskQueue;
+
+  @Getter(AccessLevel.PACKAGE)
+  @VisibleForTesting
+  private final Multimap<String, PipelineTaskQueue.TaskSpec> tasksByQueue = LinkedHashMultimap.create();
+
+  private final Collection<PipelineTaskQueue.TaskReference> taskReferences = new ArrayList<>();
+
+  public void commit() {
+    //noinspection unchecked
+    try {
+      taskReferences.addAll(this.commitTasks());
+      // returning void for simplicity, we never do anything with the response
+      dsTransaction.commit();
+    } catch (Throwable t) {
+      rollbackTasks();
```

Comment:
this case in practice shouldn't happen if what fails is the `commitTasks` step, as it should now delete anything that failed.

Reply:
I left that for the caller, to keep the call of rollbackIfActive as before.